django/utils/numberformat.py

```diff
@@ -18,46 +18,99 @@ def format(number, decimal_sep, decimal_pos=None, grouping=0, thousand_sep='',
+        grouping = 0
+
     # Make the common case fast
-    if isinstance(number, int) and not use_grouping and not decimal_pos:
-        return mark_safe(number)
-    # sign
-    sign = ''
-    # Treat potentially very large/small floats as Decimals.
-    if isinstance(number, float) and 'e' in str(number).lower():
-        number = Decimal(str(number))
+    if isinstance(number, int) and (grouping == 0 or grouping == 3):
+        return _format_int(number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    # Handle Decimals
     if isinstance(number, Decimal):
+        return _format_dec(number, decimal_sep, decimal_pos, thousand_sep, grouping)
 
-        if decimal_pos is not None:
-            # If the provided number is too small to affect any of the visible
-            # decimal places, consider it equal to '0'.
-            cutoff = Decimal('0.' + '1'.rjust(decimal_pos, '0'))
-            if abs(number) < cutoff:
-                number = Decimal('0')
-
-        # Format values with more than 200 digits (an arbitrary cutoff) using
-        # scientific notation to avoid high memory usage in {:f}'.format().
-        _, digits, exponent = number.as_tuple()
-        if abs(exponent) + len(digits) > 200:
-            number = '{:e}'.format(number)
-            coefficient, exponent = number.split('e')
-            # Format the coefficient.
-            coefficient = format(
-                coefficient, decimal_sep, decimal_pos, grouping,
-                thousand_sep, force_grouping, use_l10n,
-            )
-            return '{}e{}'.format(coefficient, exponent)
-        else:
-            str_number = '{:f}'.format(number)
+    # Don't unnecessarily convert number to string for speed
+    if isinstance(number, str):
+        return _format_string(number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    str_number = str(number)
+    # Treat potentially very large/small floats as Decimals.
+    if isinstance(number, float) and 'e' in str_number.lower():
+        return _format_dec(Decimal(str_number), decimal_sep, decimal_pos, thousand_sep, grouping)
+
+    # Format all other cases as a string
+    return _format_string(str_number, decimal_sep, decimal_pos, thousand_sep, grouping)
+
+
+def _format_int(number, decimal_sep, decimal_pos, thousand_sep, grouping):
+    """
+    Formats an integer, returning it as a string.
+
+    This function should only be used when grouping is 0 or 3. See
+    numberformat.format docstring for details about arguments.
+    """
+    if grouping and (number > 999 or number < -999):
+        # f'{number:,}' returns a string with a ',' as a thousand separator
+        number = f'{number:,}'.replace(',', thousand_sep)
     else:
-        str_number = str(number)
+        # It is unclear as to why this case should be marked safe, but
+        # we return SafeString here to avoid a change in functionality
+        # from the previous implementation of format.
+        number = SafeString(number)
```

Comment:
I don't think you necessarily need this _here_ (if at all, but I get the historic reasoning) fwiw.

In the original (if I'm reading the side-by-side correctly) it ends up marked as safe when `dec_pos` is _Falsy_. But in this changed version, it becomes a `SafeString` _before_ checking `if decimal_pos`.

If there 's _truthy_ `decimal_pos` you end up doing `SafeString` + sep (presumably str, but could be another `SafeString`) + `str` so you ultimately get back a `str` type regardless, because `'0' * decimal_pos` is the final addition and `SafeString` only defines `__add__` for another of it's kind.

If you were able to defer making it a `SafeString` until _after_ the `decimal_pos` check, you can avoid 1 or possibly 2, accounting for sep) in the _wants decimal_ case, because `str` + (potentially) `SafeString` yields `str`.

NB: that's all off the top of my head from reading the diff, caveat emptor :)